### PR TITLE
chore(refactor): refactor exit handler and tests  

### PR DESCRIPTION
**Pull request by wraithgar**
* npm mock logger writes to npm.log.record too now
 * No more extra process.exit from within the process `exit` event handle.
 * No more `exit()` function.  Logic is rolled up into the exit handler.
 * Now there is only an exit handler and an exit event listener.
   `lib/utils/perf.js` was rolled up into npm.js itself.

Unfortunately the tests were written in such a way that any further
refactoring of the exit handler was going to require also rewriting the
tests.  Fortunately NOW the tests are interacting with the exit handler
in a way that shouldn't require them to be rewritten AGAIN if we change
the internals of the exit handler.

Needs https://github.com/npm/cli/pull/3479 to land first

### CODE REVIEW

**Comment by wraithgar** [comment]
~We are either going to need to put `process.exit()` back into the exit handler or find a way to unref/cancel the update notifier promise.~

This is done, with a more meaningful comment explaining why it's in there

File: lib/base-command.js [hunk 1]
```diff
@@ -75,7 +75,6 @@ class BaseCommand {
   }
 
   async setWorkspaces (filters) {
-    // TODO npm guards workspaces/global mode so we should use this.npm.prefix?
```
Comment (nlf) [inline lib/base-command.js]: is this TODO not something we still need to do? is this an answered question already?

File: lib/npm.js [hunk 2]
```diff
@@ -37,23 +32,51 @@ const proxyCmds = new Proxy({}, {
...
   constructor () {
     super()
-    // TODO make this only ever load once (or unload) in tests
-    require('./utils/perf.js')
     this.started = Date.now()
     this.command = null
     this.commands = proxyCmds
+    this.timings = timings
+    this.timers = timers
+    this.perfStart()
```
Comment (nlf) [inline lib/npm.js:79]: this is a great intermediate step to avoid the singleton issue :tada:

File: lib/utils/exit-handler.js [hunk 3]
```diff
@@ -24,17 +21,15 @@ const getLogFile = () => {
...
-process.on('timing', (name, value) => {
-  if (timings[name])
-    timings[name] += value
-  else
-    timings[name] = value
-})
-
 process.on('exit', code => {
+  // process.emit is synchronous, so the timeEnd handler will run before the
+  // unfinished timer check below
```
Comment (nlf) [inline lib/utils/exit-handler.js:26]: :clap: thank you for calling this out, it's a super nice reminder!

**Review by nlf — APPROVED** [review]
i'm a big fan of these pull requests that delete a bunch of misdirection and just generally unnecessary code

left one question, but this all looks good as far as i can see. much simpler and much more clear :+1:

File: lib/base-command.js (hunk 1 above)
Comment (wraithgar) [reply lib/base-command.js]: No, the entire approach to npm.prefix vs npm.localPrefix etc needs addressing and I'd rather leave these comments in until we actually properly audit and review how we're using those.

**Comment by wraithgar** [comment]
rebasing fixed the conflicts w/o having to change anything. PHEW

**Pull request merged**